src/vsr.zig

```diff
@@ -1136,32 +1208,97 @@ const ViewChangeHeadersSlice = struct {
     ///   a view prior to the log_view, they would already be part of the headers.
     pub fn view_for_op(headers: ViewChangeHeadersSlice, op: u64, log_view: u32) ViewRange {
         const header_newest = &headers.slice[0];
-        const header_oldest = &headers.slice[headers.slice.len - 1];
+        const header_oldest = blk: {
+            var oldest: ?usize = null;
+            for (headers.slice) |*header, i| {
+                switch (Headers.dvc_header_type(header)) {
+                    .blank => assert(i > 0),
+                    .fault => assert(i > 0),
+                    .valid => oldest = i,
+                }
+            }
+            break :blk &headers.slice[oldest.?];
+        };
+        assert(header_oldest.op <= header_newest.op);
 
         if (op < header_oldest.op) return .{ .min = 0, .max = header_oldest.view };
         if (op > header_newest.op) return .{ .min = log_view, .max = log_view };
 
         for (headers.slice) |*header| {
-            if (header.op == op) return .{ .min = header.view, .max = header.view };
+            switch (Headers.dvc_header_type(header)) {
+                .blank => {},
+                .fault => {},
+                .valid => if (header.op == op) return .{ .min = header.view, .max = header.view },
+            }
         }
 
-        for (headers.slice[0 .. headers.slice.len - 1]) |*header_next, header_next_index| {
-            const header_prev = headers.slice[header_next_index + 1];
-            if (header_prev.op < op and op < header_next.op) {
-                return .{ .min = header_prev.view, .max = header_next.view };
+        var header_next = &headers.slice[0];
+        assert(Headers.dvc_header_type(header_next) == .valid);
+
+        for (headers.slice[1..]) |*header_prev| {
+            if (Headers.dvc_header_type(header_prev) == .valid) {
+                if (header_prev.op < op and op < header_next.op) {
+                    return .{ .min = header_prev.view, .max = header_next.view };
+                }
+                header_next = header_prev;
             }
         }
         unreachable;
     }
+
+    pub const Iterator = struct {
```

Comment:
I wonder if Iterator pulls its weight? The alternative would be something like

```zig
if (header.is_fake()) continue;                 
```

condition at the two call-sites we have.

Though, yeah, we assert here more thoroughly.

Reply:
Removed. The asserts were redundant since `verify()` should be called any time the slice is initialized.

Reply:
I think you forgot to remove/push removal of Iterator itself?

Reply:
Oops; removed.